cibyl/plugins/openstack/sources/jenkins_job_builder.py

```diff
@@ -0,0 +1,214 @@
+
+LOG = logging.getLogger(__name__)
+
+TOPOLOGY = r'[a-zA-Z0-9:,]+:[0-9]+' # controller:3,database:3,compute:2
+NODE_NAME_COUNTER = r'[a-zA-Z]+:[0-9]+'
+IP_VERSION = r'--network-protocol\s+ipv[4|6]'
+IP_VERSION_NUMBER = r'4|6'
+
+
+class JenkinsJobBuilder(SourceExtension):
+    def _parse_xml(self, path):
+        """
+        Parse xml file generated by the jenkins job builder and get all
+        groovy script sections found in the file.
+
+        param: str path: the path to the xml file
+        return: the StringIO instance (in memory file like instance) containing
+                the groovy scripts found in the generated xml file.
+        rtype: StringIO
+        """
+        root = ET.parse(path).getroot()
+        result = ""
+        for script in root.iter("script"):
+            result = result + "\n" + script.text
+        return StringIO(result)
+
+    def _get_nodes(self, path, **kwargs):
+        """
+        extract topology from the JJB xml file and
+        represent as a Nodes dictionary
+
+        Note: this function is not used to support filtering
+
+        :param path: to JJB xml file
+        :param **kwargs: cibyl command line
+
+        :return: dictionary of Nodes
+        """
+        topology = self._get_topology(path, **kwargs)
+        nodes = {}
+        for component in topology.split(","):
+            try:
+                role, amount = component.split(":")
+            except ValueError:
+                continue
+            for i in range(int(amount)):
+                node_name = role + f"-{i}"
+                nodes[node_name] = Node(node_name, role=role)
+        return nodes
+
+    def _get_topology(self, path, **kwargs):
+        """
+        extract topology from the JJB xml file and
+        represent it in the form of string, e.g
+           controller:3,compute:2
+
+        Note: this function is used to support filtering
+            e.g. --topology cont, --topology controller:3
+        :param path: to JJB xml file
+        :param **kwargs: cibyl command line
+
+        :return: topology string
+        """
+        topology_str = ""
+        if "topology" in kwargs:
+            in_mem_file = self._parse_xml(path)
+            result = set([])
+
+            lines = [line.rstrip() for line in in_mem_file if
+                     "TOPOLOGY=" in line or "TOPOLOGY =" in line]
+            for line in lines:
+                topology_lst = re.findall(TOPOLOGY, line)
+                for el in topology_lst:
+                    nodeSet = set(re.findall(NODE_NAME_COUNTER, el))
+                    result = result.union(nodeSet)
+
+            topology_str = ",".join(sorted(result))
+            # filtering support e.g. --topology cont, --topology controller:3
+            if kwargs['topology'].value is not None and len(
+                    kwargs['topology'].value) > 0 and len(
+                    list(filter(lambda x: kwargs['topology'].value[0] in x,
```

Comment:
We should use all the values passed by the user to filter, not just the first one

Reply:
> We should use all the values passed by the user to filter, not just the first one

Done